README.md
Enzyme
=======

[![Join the chat at https://gitter.im/enzymejs/enzyme](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/enzymejs/enzyme?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

[![npm Version](https://img.shields.io/npm/v/enzyme.svg)](https://www.npmjs.com/package/enzyme) [![License](https://img.shields.io/npm/l/enzyme.svg)](https://github.com/enzymejs/enzyme/blob/master/LICENSE.md) [![Build Status](https://travis-ci.org/enzymejs/enzyme.svg)](https://travis-ci.org/enzymejs/enzyme) [![Coverage Status](https://codecov.io/gh/enzymejs/enzyme/branch/master/graph/badge.svg)](https://codecov.io/gh/enzymejs/enzyme/branch/master)


Enzyme is a JavaScript Testing utility for React that makes it easier to test your React Components' output.
You can also manipulate, traverse, and in some ways simulate runtime given the output.

Enzyme's API is meant to be intuitive and flexible by mimicking jQuery's API for DOM manipulation
and traversal.

Upgrading from Enzyme 2.x or React < 16
===========

Are you here to check whether or not Enzyme is compatible with React 16? Are you currently using
Enzyme 2.x? Great! Check out our [migration guide](/docs/guides/migration-from-2-to-3.md) for help
moving on to Enzyme v3 where React 16 is supported.

### [Installation](/docs/installation/README.md)

To get started with enzyme, you can simply install it via npm. You will need to install enzyme
along with an Adapter corresponding to the version of react (or other UI Component library) you
are using. For instance, if you are using enzyme with React 16, you can run:

```bash
npm i --save-dev enzyme enzyme-adapter-react-16
```

Each adapter may have additional peer dependencies which you will need to install as well. For instance,
`enzyme-adapter-react-16` has peer dependencies on `react` and `react-dom`.

At the moment, Enzyme has adapters that provide compatibility with `React 16.x`, `React 15.x`,
`React 0.14.x` and `React 0.13.x`.

The following adapters are officially provided by enzyme, and have the following compatibility with
React:

| Enzyme Adapter Package | React semver compatibility |
| --- | --- |
| `enzyme-adapter-react-16` | `^16.4.0-0` |
| `enzyme-adapter-react-16.3` | `~16.3.0-0` |
| `enzyme-adapter-react-16.2` | `~16.2` |
| `enzyme-adapter-react-16.1` | <code>~16.0.0-0 &#124;&#124; ~16.1</code> |
| `enzyme-adapter-react-15` | `^15.5.0` |
| `enzyme-adapter-react-15.4` | `15.0.0-0 - 15.4.x` |
| `enzyme-adapter-react-14` | `^0.14.0` |
| `enzyme-adapter-react-13` | `^0.13.0` |

Finally, you need to configure enzyme to use the adapter you want it to use. To do this, you can use
the top level `configure(...)` API.

```js
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });
```

3rd Party Adapters
=============

It is possible for the community to create additional (non-official) adapters that will make enzyme
work with other libraries. If you have made one and it's not included in the list below, feel free
to make a PR to this README and add a link to it! The known 3rd party adapters are:

| Adapter Package | For Library | Status |
| --- | --- | --- |
| [`enzyme-adapter-preact-pure`](https://github.com/preactjs/enzyme-adapter-preact-pure) | [`preact`](https://github.com/developit/preact) | (stable) |
|[`enzyme-adapter-inferno`](https://github.com/bbc/enzyme-adapter-inferno)|[`inferno`](https://github.com/infernojs/inferno)|(work in progress)|

Running Enzyme Tests
===========

Enzyme is unopinionated regarding which test runner or assertion library you use, and should be
compatible with all major test runners and assertion libraries out there. The documentation and
examples for enzyme use [Mocha](https://mochajs.org) and [Chai](https://chaijs.com), but you
should be able to extrapolate to your framework of choice.

If you are interested in using enzyme with custom assertions and convenience functions for
testing your React components, you can consider using:

* [`chai-enzyme`](https://github.com/producthunt/chai-enzyme) with Mocha/Chai.
* [`jasmine-enzyme`](https://github.com/FormidableLabs/enzyme-matchers/tree/master/packages/jasmine-enzyme) with Jasmine.
* [`jest-enzyme`](https://github.com/FormidableLabs/enzyme-matchers/tree/master/packages/jest-enzyme) with Jest.
* [`should-enzyme`](https://github.com/rkotze/should-enzyme) for should.js.
* [`expect-enzyme`](https://github.com/PsychoLlama/expect-enzyme) for expect.


[Using Enzyme with Mocha](/docs/guides/mocha.md)

[Using Enzyme with Karma](/docs/guides/karma.md)

[Using Enzyme with Browserify](/docs/guides/browserify.md)

[Using Enzyme with SystemJS](/docs/guides/systemjs.md)

[Using Enzyme with Webpack](/docs/guides/webpack.md)

[Using Enzyme with JSDOM](/docs/guides/jsdom.md)

[Using Enzyme with React Native](/docs/guides/react-native.md)

[Using Enzyme with Jest](/docs/guides/jest.md)

[Using Enzyme with Lab](/docs/guides/lab.md)

[Using Enzyme with Tape and AVA](/docs/guides/tape-ava.md)

Basic Usage
===========

## [Shallow Rendering](/docs/api/shallow.md)

```javascript
import React from 'react';
import { expect } from 'chai';
import { shallow } from 'enzyme';
import sinon from 'sinon';

import MyComponent from './MyComponent';
import Foo from './Foo';

describe('<MyComponent />', () => {
  it('renders three <Foo /> components', () => {
    const wrapper = shallow(<MyComponent />);
    expect(wrapper.find(Foo)).to.have.lengthOf(3);
  });

  it('renders an `.icon-star`', () => {
    const wrapper = shallow(<MyComponent />);
    expect(wrapper.find('.icon-star')).to.have.lengthOf(1);
  });

  it('renders children when passed in', () => {
    const wrapper = shallow((
      <MyComponent>
        <div className="unique" />
      </MyComponent>
    ));
    expect(wrapper.contains(<div className="unique" />)).to.equal(true);
  });

  it('simulates click events', () => {
    const onButtonClick = sinon.spy();
    const wrapper = shallow(<Foo onButtonClick={onButtonClick} />);
    wrapper.find('button').simulate('click');
    expect(onButtonClick).to.have.property('callCount', 1);
  });
});
```

Read the full [API Documentation](/docs/api/shallow.md)



## [Full DOM Rendering](/docs/api/mount.md)

```javascript
import React from 'react';
import sinon from 'sinon';
import { expect } from 'chai';
import { mount } from 'enzyme';

import Foo from './Foo';

describe('<Foo />', () => {
  it('allows us to set props', () => {
    const wrapper = mount(<Foo bar="baz" />);
    expect(wrapper.props().bar).to.equal('baz');
    wrapper.setProps({ bar: 'foo' });
    expect(wrapper.props().bar).to.equal('foo');
  });

  it('simulates click events', () => {
    const onButtonClick = sinon.spy();
    const wrapper = mount((
      <Foo onButtonClick={onButtonClick} />
    ));
    wrapper.find('button').simulate('click');
    expect(onButtonClick).to.have.property('callCount', 1);
  });

  it('calls componentDidMount', () => {
    sinon.spy(Foo.prototype, 'componentDidMount');
    const wrapper = mount(<Foo />);
    expect(Foo.prototype.componentDidMount).to.have.property('callCount', 1);
    Foo.prototype.componentDidMount.restore();
  });
});
```

Read the full [API Documentation](/docs/api/mount.md)


## [Static Rendered Markup](/docs/api/render.md)

```javascript
import React from 'react';
import { expect } from 'chai';
import { render } from 'enzyme';

import Foo from './Foo';

describe('<Foo />', () => {
  it('renders three `.foo-bar`s', () => {
    const wrapper = render(<Foo />);
    expect(wrapper.find('.foo-bar')).to.have.lengthOf(3);
  });

  it('renders the title', () => {
    const wrapper = render(<Foo title="unique" />);
    expect(wrapper.text()).to.contain('unique');
  });
});
```

Read the full [API Documentation](/docs/api/render.md)

### React Hooks support

Enzyme supports [react hooks](https://reactjs.org/docs/hooks-intro.html) with some limitations in [`.shallow()`](https://enzymejs.github.io/enzyme/docs/api/shallow.html) due to upstream issues in React's shallow renderer:

* `useEffect()` and `useLayoutEffect()` don't get called in the React shallow renderer. [Related issue](https://github.com/facebook/react/issues/15275)

* `useCallback()` doesn't memoize callback in React shallow renderer. [Related issue](https://github.com/facebook/react/issues/15774)

#### [`ReactTestUtils.act()`](https://reactjs.org/docs/test-utils.html#act) wrap

If you're using React 16.8+ and `.mount()`, Enzyme will wrap apis including [`.simulate()`](https://enzymejs.github.io/enzyme/docs/api/ReactWrapper/simulate.html), [`.setProps()`](https://enzymejs.github.io/enzyme/docs/api/ReactWrapper/setProps.html), [`.setContext()`](https://enzymejs.github.io/enzyme/docs/api/ReactWrapper/setContext.html), [`.invoke()`](https://enzymejs.github.io/enzyme/docs/api/ReactWrapper/invoke.html) with [`ReactTestUtils.act()`](https://reactjs.org/docs/test-utils.html#act) so you don't need to manually wrap it.

A common pattern to trigger handlers with `.act()` and assert is:

```javascript
const wrapper = mount(<SomeComponent />);
act(() => wrapper.prop('handler')());
wrapper.update();
expect(/* ... */);
```

We cannot wrap the result of `.prop()` (or `.props()`) with `.act()` in Enzyme internally since it will break the equality of the returned value.
However, you could use `.invoke()` to simplify the code:

```javascript
const wrapper = mount(<SomeComponent />);
wrapper.invoke('handler')();
expect(/* ... */);
```

### Future

[Enzyme Future](/docs/future.md)


### Contributing

See the [Contributors Guide](/CONTRIBUTING.md)

### In the wild

Organizations and projects using `enzyme` can list themselves [here](INTHEWILD.md).

### License

[MIT](/LICENSE.md)


CHANGELOG.md
# Change Log

## 3.11.0

### New Stuff
 - `render`: handle Fiber strings and numbers (#2221)

### Fixes
 - `shallow`: Share child context logic between `shallow` and `dive` (#2296)
 - `mount`: `children`: include text nodes ($2269)
 - `mount`: `invoke`: use adapter’s `wrapInvoke` if present (#2158)

### Docs
 - `mount`/`shallow`: `closest`/`parent`: Add missing arguments description (#2264)
 - `mount`/`shallow`: fix pluralization of “exist” (#2262)
 - `shallow`/`mount`: `simulate`: added functional component example to simulate doc (#2248)
 - `mount`: `debug`: add missing verbose option flag (#2184)
 - `mount`/`shallow`: `update`: fix semantics description (#2194)
 - add missing backticks to linked method names (#2170)
 - `invoke`: Add missing backticks to end of codeblock (#2160)
 - `invoke`: Fix typo (#2167)
 - Explicit React CSS selector syntax description (#2178)

### Meta Stuff
 - [meta] add `funding` field
 - [meta] Update airbnb.io URLs to use https (#2222)
 - [deps] update `is-boolean-object`, `is-callable`, `is-number-object`, `is-string`, `enzyme-shallow-equal`, `array.prototype.flat`, `function.prototype.name`, `html-element-map`, `is-regex`, `object-inspect`, `object-is`, `object.entries`, `object.vales`, `raf`, `string.prototype.trim`
 - [dev deps] update `eslint`, `eslint-plugin-import`, `eslint-plugin-markdown`, `eslint-plugin-react`, `safe-publish-latest`, `eslint-config-airbnb`, `rimraf`, `safe-publish-latest`, `karma-firefox-launcher`, `babel-preset-airbnb`, `glob-gitignore`, `semver`, `eslint-plugin-jsx-a11y`

## 3.10.0

### New Stuff
 - `shallow` add `suspenseFallback` option; support `Suspense`/`Lazy` (#1975)
 - `shallow`/`mount`: add `invoke(propName)(...args)` (#1856, #945)
 - `shallow`: Support rendering and `dive()`ing `createContext()` providers and consumers (#1966)
 - `mount`/`shallow`: add `getWrappingComponent` (#1960)
 - `Utils`: add `isCustomComponent` (#1960)
 - `Utils`: add `stub` argument to `spyMethod`
 - `EnzymeAdapter`: add `matchesElementType` (#2146)
 - `RSTTraversal`: add `getHTMLFromHostNodes`

### Fixes
 - `selectors`: unwrap `memo` elements - in both directions (#2146)
 - `shallow`: properly determine "should render" for `PureComponent`s (#2096)
 - `mount`/`shallow`: `renderProp`: improve error messages (#2070)
 - `mount`: `setContext`: use proper wrapper name in error message
 - `shallow`: `.contains()`: clean up error message
 - `shallow`/`mount`: `hasClass`: avoid a crash with a non-string argument (#2057)
 - `mount`: properly handle HTML of multiple nodes (#2052)
 - `shallow`: ensure that if gDSFP exists, cDU is called. (#2027)
 - `shallow`: Mock sCU if gDSFP defined in shallow renderer rerender (#1981)
 - `mount`: `.state()`: allow getting state from stateful children of a stateless root (#2043)
 - `mount`: `.text()`: properly recurse through fragments and arrays (#2028)

### Refactors
 - `ReactWrapper`/`ShallowWrapper`: ensure calling an adapter‘s nodeToElement preserves the receiver
 - `mount`: use `getHTMLFromHostNodes`

### Docs
 - explain why need to re-call .find() after update (#2140)
 - `shallow`: fix childAt (#2134)
 - Update v2 -> v3 migration guide re props after a stage change (#1300)
 - `debug`: Added documentation for `verbose` flag (#2104)
 - Add on the fly JSDOM include example (#2072)
 - `reduce`/`reduceRight`: fix example code (#2066, #2065, #2064)
 - update `simulateError` with `getDerivedStateFromError` (#2036)
 - `shallow`: `.hasClass`: fix use of `mount`
 - add link to Cheerio API and another example (#756)
 - `jest`: Update deprecated jest config key (#2024)

## 3.9.0

### New Stuff
 - `shallow`: `isEmptyRender`: allow on multiple elements (#1924)
 - `mount`: support `:focus` selector (#1965)
 - `shallow`: Support `getChildContext()` (#1971)
 - `shallow`/`mount`: `.find`: find HTML elements by their constructor (#1933)
 - `shallow`/`mount`: `selectors`: support universal selector (#1945)
 - `mount`/`shallow`: add regex support to `.hasClass` (#1987)

### Fixes
 - `mount`/`shallow`: `.text`/`.html`: handle an array of nodes properly (#2001)
 - `shallow`: `.setProps()`: only call cDU once (#2007)
 - `mount`: ensure `findWhere` doesn‘t abort early on an empty string (#1995)
 - `mount`: `findWhere`: avoid passing empty wrappers to predicate
 - `mount`: `text()` via `findWhere`: return the string for a wrapper around a string
 - `mount`/`shallow`: Fix .exists not accepting any EnzymeSelector (#1934)
 - `mount`: Recursive `isEmptyRender` (#1924)
 - `mount`: `setState`: allow setting state on a class child of an SFC root

### Documentation
 - [jest] update Jest to setupFilesAfterEnv (#2015)
 - [jest] Change deprecated `setupTestFrameworkScriptFile` to `setupFilesAfterEnv` (#2013)
 - `mount`: `setState`: note that you can now call it on "not the root" (#2010)
 - general improvements (#1947)
 - Fix typos (#1992)
 - Added missing pages: `isEmptyRender`, `renderProp`, `equals`, SystemJS guide (#1984, #1985)
 - Fix link to .shallow() (#1951)
 - update jsdoc to use "EnzymeSelector" where applicable

### Meta Stuff
 - add "directory" field to package.json

## 3.8.0

### New Stuff
 - `shallow`/`mount`: add `renderProp` (#1863, #1891)

### Fixes
 - `shallow`/`mount`: improve error message for "single node" assertion (#1904)
 - `shallow`: shallow compare, not deep compare, state and props to determine rerendering (#1915)

### Documentation
 - `shallow`: `dive`: add a note about throwing an error (#1905)
 - `selectors: update re `displayName` (#1932)
 - `shallow`: `get`: fixed wrong `props()` usage (#1921)
 - `shallow`: `html`: shallow renders full tree (#1912)
 - Updated broken link for “.hostNodes()” in migration guide from enzyme 2 to 3 (#1909)
 - Add tape example project link (#1898)
 - `prop`: fix typo (#1883)
 - Document full support for attribute selectors (#1881)
 - Documentation update for testing React Native with jsdom (#1873)
 - Update JSDOM docs to include {request|cancel}AnimationFrame polyfills (#1867)

env.js
'use strict';

const path = require('path');
const fs = require('fs');
const spawn = require('child_process').spawn;
const rimraf = require('rimraf');
const semver = require('semver');

const promisify = (fn) => new Promise((res, rej) => {
  const done = (err, val) => (err ? rej(err) : res(val));
  fn(done);
});
const getFile = (fpath) => promisify((cb) => fs.readFile(fpath, 'utf8', cb));
// const getFiles = fpath => promisify(cb => fs.readdir(fpath, cb));
const getJSON = (fpath) => getFile(fpath).then((json) => JSON.parse(json));
const writeFile = (fpath, src) => promisify((cb) => {
  console.log('writeFile', fpath, src);
  if (process.env.DEBUG) {
    cb();
  } else {
    fs.writeFile(fpath, src, cb);
  }
});
const writeJSON = (fpath, json, pretty = false) => writeFile(
  fpath,
  (pretty ? JSON.stringify(json, null, 2) : JSON.stringify(json)) + '\n'
);
const primraf = (fpath) => promisify((cb) => {
  console.log('rimraf', fpath);
  if (process.env.DEBUG) {
    cb();
  } else {
    rimraf(fpath, cb);
  }
});
const run = (cmd, ...args) => promisify((cb) => {
  console.log(cmd + ' ' + args.join(' '));
  const child = spawn(
    process.env.DEBUG ? 'echo' : cmd,
    process.env.DEBUG ? [cmd].concat(args) : args,
    { stdio: 'inherit' }
  );
  child.on('exit', cb);
});

// This script is executed with a single argument, indicating the version of
// react and adapters etc. that we want to set ourselves up for testing.
// should be "14" for "enzyme-adapter-react-14", "15.4" for "enzyme-adapter-react-15.4", etc.
const version = process.argv[2];

// This script will do the following:
//
// 1. remove / uninstall all relevant modules
// 2. find the adapter for the passed in version
// 3. get the package.json for the adapter
// 4. add the adapter to the dev-deps for enzyme-test-suite package
// 5. call lerna bootstrap to link all the packages
// 6. install all of the package's peer deps at the top level

const root = process.cwd();

function getAdapter(reactVersion) {
  if (semver.intersects(reactVersion, '0.13.x')) {
    return '13';
  }
  if (semver.intersects(reactVersion, '0.14.x')) {
    return '14';
  }
  if (semver.intersects(reactVersion, '^15.0.0-0')) {
    if (semver.intersects(reactVersion, '>= 15.5')) {
      return '15';
    }
    return '15.4';
  }
  if (semver.intersects(reactVersion, '^16.0.0-0')) {
    if (semver.intersects(reactVersion, '>= 16.4')) {
      return '16';
    }
    if (semver.intersects(reactVersion, '~16.3')) {
      return '16.3';
    }
    if (semver.intersects(reactVersion, '~16.2')) {
      return '16.2';
    }
    if (semver.intersects(reactVersion, '~16.0 || ~16.1')) {
      return '16.1';
    }
  }
  if (semver.intersects(reactVersion, '^17.0.0')) {
    return '17';
  }
  return null;
}
const reactVersion = version < 15 ? '0.' + version : version;
const adapterVersion = process.env.ADAPTER || getAdapter(reactVersion) || version;
const adapterName = `enzyme-adapter-react-${adapterVersion}`;
const adapterPackageJsonPath = path.join(root, 'packages', adapterName, 'package.json');
const testPackageJsonPath = path.join(root, 'packages', 'enzyme-test-suite', 'package.json');

if (!fs.statSync(adapterPackageJsonPath)) {
  throw new Error('Adapter not found: "' + adapterName + '"');
}

const packagesToRemove = [
  'react',
  'react-dom',
  'react-addons-test-utils',
  'react-test-renderer',
  'create-react-class',
].map((s) => `./node_modules/${s}`);

const additionalDirsToRemove = [
];

const rmrfs = []
  .concat(packagesToRemove)
  .concat(additionalDirsToRemove);

Promise.resolve()
  .then(() => Promise.all(rmrfs.map((s) => primraf(s))))
  .then(() => run('npm', 'i'))
  .then(() => Promise.all([
    getJSON(adapterPackageJsonPath),
    getJSON(testPackageJsonPath),
  ]))
  .then(([adapterJson, testJson]) => {
    const peerDeps = adapterJson.peerDependencies;
    const installs = Object.keys(peerDeps)
      .filter((key) => !key.startsWith('enzyme'))
      .map((key) => `${key}@${key.startsWith('react') ? reactVersion : peerDeps[key]}`);

    if (process.env.RENDERER) {
      // eslint-disable-next-line no-param-reassign
      adapterJson.dependencies['react-test-renderer'] = process.env.RENDERER;
    }

    // eslint-disable-next-line no-param-reassign
    testJson.dependencies[adapterName] = adapterJson.version;

    return writeJSON(adapterPackageJsonPath, adapterJson, true).then(() => Promise.all([
      // npm install the peer deps at the root
      run('npm', 'i', '--no-save', ...installs),

      // add the adapter to the dependencies of the test suite
      writeJSON(testPackageJsonPath, testJson, true),
    ]));
  })
  .then(() => run('lerna', 'bootstrap', '--hoist=\'react*\''))
  .then(() => getJSON(testPackageJsonPath))
  .then((testJson) => {
    // now that we've lerna bootstrapped, we can remove the adapter from the
    // package.json so there is no diff
    // eslint-disable-next-line no-param-reassign
    delete testJson.dependencies[adapterName];
    return writeJSON(testPackageJsonPath, testJson, true);
  })
  .catch((err) => console.error(err));
